### PR TITLE
Fix 2D smooth visibility bug by utilizing SpectralExtraction linking logic

### LINKED ISSUE #1956: Specviz2d: smoothed spectrum not showing
**Reporter: Camilla Pacifici**

Load a s2d file.
Use gaussian smooth plugin on 1d spectrum.
Smoothed spectrum is in the list of data, but does not show in the viewer.

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

Big thanks to Jesse and Kyle for the hack session that found this solution! This PR fixes #1956 by borrowing the linking logic from the Spectral Extraction plugin for any non-cube linking scenarios.

Turns out the fix is *slightly* more complicated than the one we figured out together; the linking logic for Spectral Extraction doesn't work for smoothed data being linked against cubes (mainly in the case of Cubeviz). Because Cubeviz wasn't impacted by this bug to begin with, and the observation that the Spectral Extraction linking logic works for Specviz too, I preserved the original logic for gaussian smooth linking in Cubeviz, while using the Spectral Extraction linking logic for the `ndim < 3` usecase.

Fixes #1956

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by duytnguyendtn** [comment]
Hmm... Test failed because the Gaussian Smooth plugin couldn't find the autocollapsed 1D spectrum. It's passing locally for me. Can anyone reproduce?

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (pllim) [inline jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:185]: There are only 10 data points. I feel safer if you test specifically for the 10 expected values.

Also, `mark.x` is the spectral axis, right? It should follow the input spectral axis?

Might also want to use `np.testing.assert_allclose` instead of `assert np.<something>`.

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py:230]: So the reason why I left these args is two thoughts:

1. If you look above, I implemented this for the cube fixture because one test I added specifically needed the cube in a different unit. And rather than copying a whole new fixture, I just added unit args. So there's a precedent for it
2. Because of the upcoming unit conversion work, combined with the uneven wavelength axes work coming up soon, I figure it will be useful in those tests. 🤷

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py]: Good point! Implemented in https://github.com/spacetelescope/jdaviz/pull/2023/commits/6179460e4107d851f48e5c31ab6161f90d624a67.

I couldn't do it for `mark.y` because the marks are different. I'm testing it in a loop rather than each element directly because I'm trying to avoid some javascript error I got when accessing `marks[n]` that didn't show up when I called `for mark in marks`. Strange...

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (duytnguyendtn) [reply jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:173]: 100 was my test value when doing this by hand so I could see the difference, and I just copied it over. I didn't notice a change in computation time between 1 and 100, but let me know if you strongly want this changed

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py]: Good point! Implemented in https://github.com/spacetelescope/jdaviz/pull/2023/commits/e8fbab40fd7258ddaa52bc91ef93d71f5df04225

File: jdaviz/conftest.py
Comment (pllim) [reply jdaviz/conftest.py:230]: I actually don't think you can define a fixture like that, at least not from the examples I see in https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#fixture-parametrize . Is there a working example from somewhere you are basing this implementation from?

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (pllim) [reply jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:173]: Probably ok here, just looks a little jarring.

**Comment by pllim** [comment]
> The autocollapsed spectrum is supposed to generate automatically. Why wouldn't it be available?

I don't understand that part either but I cannot access the collapsed spectrum in Cubeviz from `data_collection`. I don't know why you don't see that failure though!

**Comment by duytnguyendtn** [comment]
And the mystery continues @pllim! Windows and OSX CI is now passing 😱 ! The failure on 3.10 is due to an SSL error

**Comment by pllim** [comment]
Tsk tsk, the failure is from astroquery, see #2021

**Comment by pllim** [comment]
> Windows and OSX CI is now passing

🤯  Still failing for me locally.

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py:230]: Yep! You can see it in action on main here: https://github.com/spacetelescope/jdaviz/blob/f8b31919df8ff91d3db90c9a644b5a276a53295f/jdaviz/configs/specviz/plugins/line_analysis/tests/test_lineflux.py#L77

I'm basically parameterizing a bunch of different test cases I define higher up and "on the fly" generate different cubes of different flux units: https://github.com/spacetelescope/jdaviz/blob/f8b31919df8ff91d3db90c9a644b5a276a53295f/jdaviz/configs/specviz/plugins/line_analysis/tests/test_lineflux.py#L13-L21

This also, separately, allows me to generate new pytest fixtures dynamically and easily without having to duplicate code that all looks VERY similar: https://github.com/spacetelescope/jdaviz/blob/f8b31919df8ff91d3db90c9a644b5a276a53295f/jdaviz/conftest.py#L182-L206

**Comment by javerbukh** [comment]
I'm in a fresh conda environment and I'm still unable to see the smoothed spectrum in specviz2d. Is that only me?

**Comment by kecnry** [comment]
> I'm in a fresh conda environment and I'm still unable to see the smoothed spectrum in specviz2d. Is that only me?

Make sure you're selecting the 1d spectrum as the input.  There is a separate ticket for fixing the 2d spectral smoothing (or maybe for now we should disable the 2D spectrum as input).

**Comment by pllim** [comment]
Okay, so I know why it fails for me now but I dunno why it doesn't fail for everyone else.

I get `UserWarning: background window extends beyond image boundaries (3.5 >= 3)` if I raise an error where specviz2d helper emits "Automatic spectrum extraction failed" message.

**Comment by duytnguyendtn** [comment]
Ahh @pllim I increased the default 2D image size to try and avoid that warning in https://github.com/spacetelescope/jdaviz/pull/2023/commits/e8fbab40fd7258ddaa52bc91ef93d71f5df04225. Try testing on that commit if you aren't already?

**Comment by pllim** [comment]
@duytnguyendtn , it passes locally for me now. Thanks!

File: jdaviz/conftest.py
Comment (pllim) [reply jdaviz/conftest.py:230]: Still not sure about those args in this fixture though.

**Comment by pllim** [comment]
BTW, CI should pass now if you rebase. FYI.

**Comment by javerbukh** [comment]
That worked for me @kecnry thanks! I can approve once CI is green.

**Comment by pllim** [comment]
I still think that pytest fixture should not have any input argument BTW.

**Review by kecnry — APPROVED** [review]
Approving pending the fixture conversation being resolved with @pllim.  Thanks for tracking this down!

~Does need change log entry and rebasing/passing CI before merge though!~

File: jdaviz/conftest.py
Comment (pllim) [inline jdaviz/conftest.py:238]:
```suggestion
    return Spectrum1D(flux=data*u.MJy, spectral_axis=data[3]*u.um)
```

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (pllim) [inline jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:187]:
```suggestion
        assert not np.array_equal(mark.y, [0, 1])
```

File: jdaviz/conftest.py
Comment (duytnguyendtn) [reply jdaviz/conftest.py]: Fair point! I removed it in https://github.com/spacetelescope/jdaviz/pull/2023/commits/c2ce1007401cf7c6bd3f3e06fffea166f45a5656 and we can revisit if it ever becomes useful in the future

**Review by pllim — APPROVED** [review]
LGTM now, though one remaining nitpick on the name branding? Is it uppercase D or lowercase D?

File: CHANGES.rst
Comment (duytnguyendtn) [reply CHANGES.rst:76]: Docs have it as capital D: https://jdaviz.readthedocs.io/en/latest/specviz2d/index.html

Comment (pllim) [reply CHANGES.rst:76]: Oh, weird. I thought I remember we agreed to rebrand so that only first letter is capitalized (when we dropped the capital V). Also the class name has lowercase "d".

https://github.com/spacetelescope/jdaviz/blob/71a514c1cf7741874d2770f34c5d0ff9045038a7/jdaviz/configs/specviz2d/helper.py#L12

Now I am confused. But I guess this discussion should not block merge. I already approved. Thanks for your patience!

**Comment by duytnguyendtn** [comment]
Failure is due to unrelated `asdf_extension` warning in `specutils`. Merging after consultation with the team. Thanks all!

**Pull request merged**
